src/librustc_typeck/check/wfcheck.rs

```diff
@@ -354,19 +343,100 @@ impl<'a, 'gcx> CheckTypeWellFormedVisitor<'a, 'gcx> {
                 }
             }
 
-            let predicates = fcx.tcx.predicates_of(item_def_id).instantiate_identity(fcx.tcx);
-            let predicates = fcx.normalize_associated_types_in(item.span, &predicates);
-            this.check_where_clauses(fcx, item.span, &predicates);
+            this.check_where_clauses(fcx, item.span, item_def_id);
 
             fcx.impl_implied_bounds(item_def_id, item.span)
         });
     }
 
+    /// Checks where clauses and inline bounds that are declared on def_id.
     fn check_where_clauses<'fcx, 'tcx>(&mut self,
                                        fcx: &FnCtxt<'fcx, 'gcx, 'tcx>,
                                        span: Span,
-                                       predicates: &ty::InstantiatedPredicates<'tcx>)
+                                       def_id: DefId) {
+        self.inner_check_where_clauses(fcx, span, def_id, false)
+    }
+
+    fn check_trait_where_clauses<'fcx, 'tcx>(&mut self,
+                                       fcx: &FnCtxt<'fcx, 'gcx, 'tcx>,
+                                       span: Span,
+                                       def_id: DefId) {
+        self.inner_check_where_clauses(fcx, span, def_id, true)
+    }
+
+    /// Checks where clauses and inline bounds that are declared on def_id.
+    fn inner_check_where_clauses<'fcx, 'tcx>(&mut self,
+                                       fcx: &FnCtxt<'fcx, 'gcx, 'tcx>,
+                                       span: Span,
+                                       def_id: DefId,
+                                       is_trait: bool)
     {
+        use ty::subst::Subst;
+        use rustc::ty::TypeFoldable;
+
+        let mut predicates = fcx.tcx.predicates_of(def_id);
+        let mut substituted_predicates = Vec::new();
+
+        let generics = self.tcx.generics_of(def_id);
+        let defaulted_params = generics.types.iter()
+                                             .filter(|def| def.has_default &&
+                                                     def.index >= generics.parent_count() as u32);
+        for param_def in defaulted_params {
+            // Defaults must be well-formed.
```

Comment:
Nit: I always think it helps to see concrete examples. So here we might extend the comment to say:

---

This parameter has a default value. Check that this default value is well-formed. This would rule out something like this (see test `XXX.rs`):

```
struct Foo<T = Vec<[u32]>> { .. }
// Vec<[u32]> is not WF because `[u32]: Sized` does not hold.
```